migrations/1613348381637-replaceFileLinks.js

```diff
@@ -0,0 +1,159 @@
+/* eslint-disable no-await-in-loop */
+const mongoose = require('mongoose');
+// eslint-disable-next-line no-unused-vars
+const { alert, error } = require('../src/logger');
+
+const { connect, close } = require('../src/utils/database');
+
+const { lessonSchema } = require('../src/services/lesson/model');
+const { linkSchema } = require('../src/services/link/link-model');
+
+// important lesson(s) and link(s) must write with s, because they are feather services that add the s
+const LessonModel = mongoose.model('lesson613348381637', lessonSchema, 'lessons');
+const LinkModelOld = mongoose.model('linkOld613348381637', linkSchema, 'linksBackup');
+const LinkModelNew = mongoose.model('linkNew613348381637', linkSchema, 'links');
+
+module.exports = {
+	up: async function up() {
+		await connect();
+		alert(`Migrations replaceFileLinks is starting...`);
+		// Use regex to get only matched fields
+		// /files/fileModel/
+		// [A-Za-z0-9]{10} ==> matched CUcACAnvKd
+		// example https://hpi-schul-cloud.de/link/CUcACAnvKd
+
+		const linkQuery = { target: { $regex: /files\/fileModel/g } };
+		const amountProm = LinkModelOld.find(linkQuery).countDocuments();
+
+		const $regex = /link\/[A-Za-z0-9]{10}/gm;
+		const lessonsProm = LessonModel.find({ 'contents.content.text': { $regex } })
+			.select({ 'contents.content': 1, _id: 1 })
+			.lean()
+			.exec();
+
+		const [amount, lessons] = await Promise.all([amountProm, lessonsProm]);
+
+		alert(`Found ${lessons.length} lessons with included file links.`);
+		alert(`Found ${amount} total fileModel links.`);
+
+		const limit = 500;
+		let looped = 0;
+		const deletedLinkTasks = [];
+
+		while (looped < amount) {
+			try {
+				alert(`start loop ${looped} < ${amount}`);
+				const links = await LinkModelOld.find(linkQuery)
+					.sort({
+						updatedAt: 1,
+						createdAt: 1,
+					})
+					.skip(looped)
+					.limit(limit)
+					.select({ _id: 1, target: 1 })
+					.lean()
+					.exec();
+
+				const machtedLinks = [];
+				// match lessons
+				// eslint-disable-next-line no-loop-func
+				links.forEach((link) => {
+					const matches = [];
+					const linkId = link._id;
+
+					lessons.forEach((lesson) => {
+						lesson.contents.forEach((entry) => {
+							if (entry.content && entry.content.text && entry.content.text.indexOf(linkId) !== -1) {
+								entry.IS_MODIFIED = true;
+								matches.push(entry);
+							}
+						});
+					});
+
+					if (matches.length > 0) {
+						link.matches = matches;
+						machtedLinks.push(link);
+					}
+				});
+
+				alert(`Effected Links: ${machtedLinks.length}`);
+
+				// all matched must replace with new entry
+				// create new link > replace link in lessons > delete old link
+				const createNewLinksPromise = [];
+				machtedLinks.forEach((link) => {
+					const prom = LinkModelNew.create({
+						target: link.target,
+						systemNote: { ticket: 'VOR-3', note: 'migration - new created', oldId: link._id },
```

Comment:
Instead of setting the old link here, we can create a hash map that maps from the old link to the new link (or the other way round). Thus, we have to store more entries in the RAM, but the matching will be faster, because we no longer need the loop below.